replacer_test.go

```diff
@@ -0,0 +1,275 @@
+	length := len(rep.static)
+	if len(rep.static) != 7 {
+		t.Errorf("Expectd length '%v' got '%v'", 7, length)
+	}
+}
+
+func TestReplacerReplaceAll(t *testing.T) {
+	rep := fakeReplacerFilled()
+
+	for _, tc := range []struct {
+		testInput string
+		expected  string
+	}{
+		{
+			// test vars without space
+			testInput: "{test1}{asdf}{1}",
+			expected:  "val1123test-123",
+		},
+		{
+			// test vars with space
+			testInput: "{test1} {asdf} {1} ",
+			expected:  "val1 123 test-123 ",
+		},
+		{
+			// test with empty val
+			testInput: "{test1} {testEmpty} {asdf} {1} ",
+			expected:  "val1 EMPTY 123 test-123 ",
+		},
+		{
+			// test vars with not finished placeholders
+			testInput: "{te{test1}{as{{df{1}",
+			expected:  "{teval1{as{{dftest-123",
+		},
+		{
+			// test with non existing vars
+			testInput: "{test1} {nope} {1} ",
+			expected:  "val1 {nope} test-123 ",
+		},
+	} {
+		actual := rep.ReplaceAll(tc.testInput, "EMPTY")
+
+		// test if all are replaced as expected
+		if actual != tc.expected {
+			t.Errorf("Expectd '%s' got '%s' for '%s'", tc.expected, actual, tc.testInput)
+		}
+	}
+}
+
+func TestReplacerReplaceAllDefaults(t *testing.T) {
+	rep := NewReplacer()
+	testInput := "{system.hostname} {system.slash} {system.os} {system.arch}"
+	expected := getHostnameForTest() + " " + string(filepath.Separator) + " " + runtime.GOOS + " " + runtime.GOARCH
+
+	// test env.
+	testInput += " {env.GOPATH}"
+	if env, ok := getEnvForTest("GOPATH"); ok {
+		expected += " " + env
+	} else {
+		expected += " "
+	}
+
+	actual := rep.ReplaceAll(testInput, "EMPTY")
+
+	// test if all are replaced as expected
+	if actual != expected {
+		t.Errorf("Expectd '%s' got '%s' for '%s'", expected, actual, testInput)
+	}
+}
+
+func TestReplacerDelete(t *testing.T) {
+	rep := fakeReplacerFilledStatic()
+	toDel := []string{
+		"key2", "key4",
+	}
+
+	for _, key := range toDel {
+		rep.Delete(key)
+
+		// test if key is removed from static map
+		if _, ok := rep.static[key]; ok {
+			t.Errorf("Expectd '%s' to be removed. It is still in static map.", key)
+		}
+	}
+
+	// check if static slice is smaller
+	expected := len(fakeReplacerFilledStatic().static) - len(toDel)
+	actual := len(rep.static)
+	if len(rep.static) != expected {
+		t.Errorf("Expectd length '%v' got lenth '%v'", expected, actual)
+	}
+}
+
+func TestReplacerMap(t *testing.T) {
+	rep := fakeReplacer()
+
+	for i, tc := range []struct {
+		key   string
+		value string
+	}{
+		{
+			key:   "f1",
+			value: "v1",
+		},
+		{
+			key:   "f2",
+			value: "v2",
+		},
+	} {
+		rep.Map(func(key string) (val string, ok bool) {
+			if key == tc.key {
+				return tc.value, true
+			}
+			return "NO", false
+		})
+
+		// test if function (which listens on specific key) is added bychecking length
+		if len(rep.providers) == i+1 {
+			val, _ := rep.providers[i](tc.key) // never fails, as we just checked the length
+			// check if the last function is the one we just added
+			if val != tc.value {
```

Comment:
A better way to do this is to just compare the pointers, i.e. `rep.providers[i] == testCaseProviderFunc` (where testCaseProviderFunc is the function you passed into `Map()`).

Reply:
It doesn't work how you suggested it. I now use Sprintf to get the pointer value like it is suggested here: https://stackoverflow.com/a/34901677

Reply:
Ah, hmm. TIL. Good idea. That'll do.